src/core/tsi/ssl/key_logging/ssl_key_logging.cc

```diff
@@ -0,0 +1,158 @@
+
+#include <map>
+
+#include "src/core/lib/gprpp/sync.h"
+#include "src/core/lib/iomgr/error.h"
+#include "src/core/lib/slice/slice_internal.h"
+#include "src/core/tsi/ssl/key_logging/ssl_key_logging.h"
+
+#include <grpc/support/log.h>
+#include <grpc/support/string_util.h>
+
+namespace tsi {
+
+#if OPENSSL_VERSION_NUMBER >= 0x10100000 && !defined(LIBRESSL_VERSION_NUMBER)
+grpc_core::Mutex* g_tls_key_logger_registry_mu;
+std::map<std::string, TlsKeyLogFileWriter*> g_tls_key_log_file_writer_map
+  ABSL_GUARDED_BY(g_tls_key_logger_registry_mu);
+static std::atomic<bool> g_tls_key_logger_registry_initialized(false);
+#endif
+
+TlsKeyLogFileWriter::TlsKeyLogFileWriter(
+    const std::string& tls_key_log_file_path)
+    : fd_(nullptr), tls_key_log_file_path_(tls_key_log_file_path) {
+  GPR_ASSERT(!tls_key_log_file_path_.empty());
+  fd_ = fopen(tls_key_log_file_path_.c_str(), "w+");
+  grpc_error_handle error = GRPC_ERROR_NONE;
+
+  if (fd_ == nullptr) {
+    error = GRPC_OS_ERROR(errno, "fopen");
+  }
+
+  if (error != GRPC_ERROR_NONE) {
+    gpr_log(GPR_ERROR,
+            "Ignoring TLS Key logging. ERROR Opening TLS Keylog "
+            "file: %s", grpc_error_std_string(error).c_str());
+  }
+};
+
+TlsKeyLogFileWriter::~TlsKeyLogFileWriter() {
+  if (fd_ != nullptr) fclose(fd_);
+  {
+    grpc_core::MutexLock lock(g_tls_key_logger_registry_mu);
+    g_tls_key_log_file_writer_map.erase(tls_key_log_file_path_);
+  }
+}
+
+void TlsKeyLogFileWriter::AppendSessionKeys(
+    SSL_CTX* /* ssl_context */,
+    const TsiTlsKeyLoggerConfig& tsi_tls_key_log_config,
+    const std::string& session_keys_info) {
+  grpc_core::MutexLock lock(&lock_);
+
+  if (fd_ == nullptr || session_keys_info.empty()) return;
+
+  switch (tsi_tls_key_log_config.key_logging_format) {
+    case GRPC_TLS_KEY_LOG_FORMAT_NSS:
+      // Append to key log file under lock
+      bool err;
+
+      err = (fwrite(session_keys_info.c_str(), sizeof(char),
+                    session_keys_info.length(),
+                    fd_) < session_keys_info.length());
+
+      // Append new-line character as well. Doing a separate syscall to avoid
+      // dynamic memory allocation and modification of passed session_keys_info
+      err |= (fwrite("\n", sizeof(char), 1, fd_) != 1);
+
+      if (err) {
+        grpc_error_handle error = GRPC_OS_ERROR(errno, "fwrite");
+        gpr_log(GPR_ERROR, "Error Appending to TLS Keylog file: %s",
+                grpc_error_std_string(error).c_str());
+        fclose(fd_);
+        fd_ = nullptr;  // disable future attempts to write to this file
+      } else {
+        fflush(fd_);
+      }
+      break;
+
+    // For future extensions to support more/custom logging formats
+    default:
+      break;
+  }
+}
+
+grpc_core::RefCountedPtr<TlsKeyLogger>
+TlsKeyLoggerRegistry::CreateTlsKeyLogger(
+    const TsiTlsKeyLoggerConfig& tsi_tls_key_log_config) {
+#if OPENSSL_VERSION_NUMBER >= 0x10100000 && !defined(LIBRESSL_VERSION_NUMBER)
```

Comment:
What was the rationale for applying these constraints? I saw in some part of our code we have this, while in some others we don't.
I think logically speaking, we only need these constraints once, right? Something like:


Can we put this in `src/core/tsi/ssl_transport_security.cc`?

Reply:
These macros are moved to ssl_transport_security.cc